### PR TITLE
Tags via custom HTTP header - provide some authentication

### PR DESCRIPTION
**Pull request by pladen**
Hi there,

Here is a PR we're using on a multi-tenant TSDB cluster (2.2.1). It allows to give tags to TSDB API via custom HTTP headers.

It basically allow to authenticate HTTP client on reverse proxy (nginx etc) with various mechanisms (SSL cert, IP address, basic auth, even SAML) and to push this info to TSDB within a custom header. This aims to securely separate data between our customers, given that the reverse proxy will properly deal with authentication.

You have to put the name of the header in your config, such as :
`tsd.http.header_tag = X-TSDB-TAG`

Then if you push data via the /api/put endpoint, just add key/value tags in this header :
`X-TSDB-TAG: customer=john,dc=virginia`

These tags and values will be added to the tags already present in JSON payload.
Example with haproxy and SSL client auth :
`http-request set-header X-TSDB-TAG customer=ssl_c_s_dn(cn)`

I am aware that some people are currently working on a real authentication framework (see #667). This patch does not aim to deal with user authentication but to permit customer isolation via third party authentication (ie: http reverse proxy).

We can write some docs about this, given that this PR is on the right track.

Note1 : tags in headers only work on "put" endpoint. It should be easy to use it also in "query" endpoint
Note2 : this PR is against next, but you can find a commit against master on our fork

Regards

### CODE REVIEW

File: src/tsd/PutDataPointRpc.java [hunk 1]
```diff
@@ -292,7 +293,18 @@ public void execute(final TSDB tsdb, final HttpQuery query)
     int queued = 0;
     final List<Deferred<Boolean>> deferreds = synchronous ? 
         new ArrayList<Deferred<Boolean>>(dps.size()) : null;
-    
+
+    if (tsdb.getConfig().enable_header_tag()) {
```
Comment (johann8384) [inline src/tsd/PutDataPointRpc.java]: For consistency, please use this method for checking the config value: 

 tsdb.getConfig().getString("tsd.http.enable_header_tag").equals("true");

File: src/utils/Config.java [hunk 2]
```diff
@@ -564,6 +577,7 @@ protected void setDefaults() {
     default_map.put("tsd.core.stats_with_port", "false");    
     default_map.put("tsd.http.show_stack_trace", "true");
     default_map.put("tsd.http.query.allow_delete", "false");
+    default_map.put("tsd.http.header_tag", "");
```
Comment (johann8384) [inline src/utils/Config.java]: Please use: default_map.put("tsd.http.header_tag", "false");

Comment (pladen) [inline src/utils/Config.java]: The "header_tag" config parameter is actually a string, representing the name of the HTTP header to look for.
ie `tsd.http.header_tag  = X-TSDB-TAG`

Do you suggest to add a header_tag_enable boolean config parameter to enable/disable this behaviour ?

File: src/tsd/PutDataPointRpc.java (hunk 1 above)
Comment (pladen) [inline src/tsd/PutDataPointRpc.java]: There is actually no "tsd.http.enable_header_tag" config parameter. See other portion of patch, i add a method to check the !null of "tsd.http.header_tag" value.

Maybe i should check this is the constructor, like how "send_telnet_errors" is managed ?

**Comment by manolama** [comment]
I think it's ok, we'd need some UTs around it as well please.

**Comment by johann8384** [comment]
I'll look at the issue with #1003 and merge this there once it's sorted.

**Pull request closed without merging**